### PR TITLE
Support for Rundeck 2.7.x

### PR DESCRIPTION
**Pull request by nilroy**
It adds support to rundeck 2.7.x

### CODE REVIEW

**Comment by nilroy** [comment]
@pcross616 @atheiman I closed my previous PR and did the changes in this PR. Now all tests are passing. Request your review.

**Comment by atheiman** [comment]
could you bump the minor version to 4.1.0?

I'll try to find time to look through this PR today or tonight.

**Comment by nilroy** [comment]
@atheiman bumped up the minor version and also updated the changelog

**Comment by nilroy** [comment]
@atheiman did you managed to review this?

File: recipes/server_install.rb [hunk 1]
```diff
@@ -265,6 +282,3 @@
     end
   end
 end
-
-include_recipe 'rundeck::_connect_rundeck_api_client'
```
Comment (atheiman) [inline recipes/server_install.rb]: why move this to the server recipe?

File: recipes/server_install.rb [hunk 2]
```diff
@@ -182,6 +188,17 @@
   notifies (node['rundeck']['restart_on_config_change'] ? :restart : :nothing), 'service[rundeck]', :delayed
 end
 
+template "/etc/init/rundeckd.conf" do
```
Comment (atheiman) [inline recipes/server_install.rb]: does this fix the ubuntu problems we've been having? if so, take the ubuntu tests out of the acceptable failures in the .travis.yml

File: recipes/server_install.rb [hunk 3]
```diff
@@ -28,6 +28,8 @@
 
 case node['platform_family']
 when 'rhel'
+  yum_package 'which'
```
Comment (atheiman) [inline recipes/server_install.rb]: why did you need to add this, and why did you add it to rhel only?

File: recipes/server_install.rb [hunk 4]
```diff
@@ -59,6 +61,8 @@
 
   rundeck_version = node['rundeck']['deb']['package'].split('-')[1]
 
+  package 'uuid-runtime'
```
Comment (atheiman) [inline recipes/server_install.rb]: why did you need to add this?

File: spec/recipes/server_install_spec.rb [hunk 5]
```diff
@@ -3,8 +3,7 @@
 describe 'rundeck::server_install' do
   # ordered list of included recipes
   let(:included_recipes) do
-    %w(rundeck::default rundeck::_data_bags rundeck::_connect_rundeck_api_client
-       rundeck::_projects)
+    %w(rundeck::default rundeck::_data_bags)
   end
```
Comment (atheiman) [inline spec/recipes/server_install_spec.rb]: you need to add specs for the other changes you made to the `server_install` recipe

File: templates/default/jaas-activedirectory.conf.erb [hunk 6]
```diff
@@ -29,7 +29,11 @@ activedirectory {
 <% end -%>
     reportStatistics="<%=@ldap[:reportstatistics]%>";
 
-    org.eclipse.jetty.plus.jaas.spi.PropertyFileLoginModule required
+    <% if Gem::Requirement.new('< 2.7.0').satisfied_by?(Gem::Version.new(@rundeck_version)) %>
```
Comment (atheiman) [inline templates/default/jaas-activedirectory.conf.erb]: I dont think `@rundeck_version` is available in this template - https://github.com/Webtrends/rundeck/blob/master/recipes/server_install.rb#L155-L160

can you spec this out and prove that the file contains `org.eclipse.jetty.plus.jaas.spi.PropertyFileLoginModule sufficient` or `org.eclipse.jetty.jaas.spi.PropertyFileLoginModule sufficient` when its supposed to? That should be easy with ChefSpec:

```ruby
expect(chef_run).to render_file('/etc/rundeck/jaas-activedirectory.conf').with_content('the content')
```

File: templates/default/profile.erb [hunk 7]
```diff
@@ -1,48 +1,59 @@
-RDECK_HOME=<%=@rundeck[:basedir]%>
-export RDECK_HOME
-
-RDECK_BASE=<%=@rundeck[:basedir]%>
-export RDECK_BASE
-
-JAVA_CMD=java
-RUNDECK_TEMPDIR=<%=@rundeck[:tempdir]%>
-
+RDECK_BASE="<%=@rundeck[:basedir]%>"
```
Comment (atheiman) [inline templates/default/profile.erb]: youre going to have to explain all these changes. I don't think anyone knows what this file does or is supposed to do so I'm not very comfortable with you changing it.

File: templates/default/rundeck-upstart.conf.erb [hunk 8]
```diff
@@ -0,0 +1,29 @@
+# RunDeck
```
Comment (atheiman) [inline templates/default/rundeck-upstart.conf.erb]: since this doesnt have any template logic in it, can you change it to `cookbook_file` resource and move this file to the `files/default` directory

File: templates/default/web.xml.erb [hunk 9]
```diff
@@ -1,269 +1,269 @@
...
+        <init-param>
+            <param-name>targetBeanName</param-name>
+            <param-value>characterEncodingFilter</param-value>
+        </init-param>
+        <init-param>
+            <param-name>targetFilterLifecycle</param-name>
+            <param-value>true</param-value>
+        </init-param>
+    </filter>
+    <% if Gem::Requirement.new('> 2.6.9').satisfied_by?(Gem::Version.new(@rundeck_version))%>
```
Comment (atheiman) [inline templates/default/web.xml.erb]: again you'll need to spec out the differences in content get rendered as expected.

also, should it be `> 2.6.9`? Wouldnt `>= 2.7.0` be more appropriate? What if 2.6.10 is released with a bug fix - I would expect you'd want the 2.6.x style of config.

File: recipes/server_install.rb (hunk 3 above)
Comment (nilroy) [inline recipes/server_install.rb]: Without this package the centos7 rundeck could not find a config file. Thats why the tests were failing for centos7

File: recipes/server_install.rb (hunk 4 above)
Comment (nilroy) [inline recipes/server_install.rb]: Without this package the ubuntu was not able to install rundeck 2.7.x

File: recipes/server_install.rb (hunk 1 above)
Comment (nilroy) [inline recipes/server_install.rb]: I think we shloud keep the server_install recipie just to install the rundeck server. I am not using the project creation from cookbooks. So moved those parts to server recipie.

File: recipes/server_install.rb (hunk 2 above)
Comment (nilroy) [inline recipes/server_install.rb]: The rundeck upstart config that comes with rundeck had a bug (https://github.com/rundeck/rundeck/pull/2246) . Thats why I added the rundeck upstart config. But it seems they fixed it in latest rundeck version. I shall remove that.

File: templates/default/jaas-activedirectory.conf.erb (hunk 6 above)
Comment (nilroy) [inline templates/default/jaas-activedirectory.conf.erb]: @rundeck_version is available in the template. You are seeing the wrong file. check this out https://github.com/nilroy/rundeck/blob/2018dfc7219cd187109fca9f3e21dfd1ae4c4413/recipes/server_install.rb#L164

File: templates/default/rundeck-upstart.conf.erb (hunk 8 above)
Comment (nilroy) [inline templates/default/rundeck-upstart.conf.erb]: It has a template logic in it. Check the line https://github.com/nilroy/rundeck/blob/2018dfc7219cd187109fca9f3e21dfd1ae4c4413/templates/default/rundeck-upstart.conf.erb#L26

File: templates/default/web.xml.erb (hunk 9 above)
Comment (nilroy) [inline templates/default/web.xml.erb]: I have tested with all three versions of 2.6.9, 2.6.10 and 2.6.11. All these versions needs this to be able to start rundeck. And they are following the same trend in 2.7.x. So I would say its the correct one.

File: templates/default/profile.erb (hunk 7 above)
Comment (nilroy) [inline templates/default/profile.erb]: I followed the rundeck upgrade docs http://rundeck.org/docs/upgrading/index.html. If you install rundeck manually you would get the profile file same as the one I put as a template. The changes that are required on use both 2.6.x (2.6.9 and onwards) and 2.7.x are included in the template.

File: templates/default/web.xml.erb (hunk 9 above)
Comment (nilroy) [inline templates/default/web.xml.erb]: Added specs in my latest commit

File: templates/default/profile.erb (hunk 7 above)
Comment (nilroy) [inline templates/default/profile.erb]: Added specs in my latest commit

File: spec/recipes/server_install_spec.rb (hunk 5 above)
Comment (nilroy) [inline spec/recipes/server_install_spec.rb]: Added specs in my latest commit

File: templates/default/jaas-activedirectory.conf.erb (hunk 6 above)
Comment (nilroy) [inline templates/default/jaas-activedirectory.conf.erb]: Added specs in my latest commit

**Comment by nilroy** [comment]
@atheiman the kitchen tests are failing from a differerent part. I smell something fishy in travis. I could run .travis.sh script locally in my development environment and tests are passing there

**Comment by nilroy** [comment]
The apache is having some issue. It could not find the init script. That's the error its showing in build. And its clearly to do with centOS. Ubuntu does not show up this error. I would assume it a temporary glith in centOS yum repo for apache or something. @atheiman any Idea why the apache is missing init script. I have not done any changes which can break those.

**Comment by nilroy** [comment]
@atheiman @pcross616 

I created another PR that uses the current master and I am seeing the same issue in travis build. https://github.com/Webtrends/rundeck/pull/155. I have no idea why its happening. Maybe some CentOS expert should intervene.

**Comment by nilroy** [comment]
@atheiman @pcross616
This PR was passing test previously. There is some issue currently with the tests in Travis. Can we go ahead with it unless you have any further comment. It's been almost 2 months I am working on this and would like to see this merged.

**Comment by nilroy** [comment]
@atheiman could you reply to my comments? It's been a while now. If you have some more comments please give those. I hope you could understand that the current tests are failing because of some issue in the Travis. If you have doubt try to test the master. If you have no further comment could you close your review comments and merge this PR? 

@pcross616 please let me know your view.

**Comment by pcross616** [comment]
i am ok with getting this in and working the builds at the same time.

**Comment by nilroy** [comment]
@pcross616 thanks a lot.
